### DISCUSSION TITLE
Unused branch in fragment shader causing excessive slowdown

### DISCUSSION THREAD

**Discussion by kpeeters** [google/filament#7377] [General]
I have a fragment shader of the form

```
fragment {
   void material(inout MaterialInputs material) {
      prepareMaterial(material);

      if( [some condition involving only uniforms] ) {
          [ slow code with loads of texture lookups ]
      }
      else {
          [ fast code ]
      }
   }
}
```

I am looking at a case in which I set the uniforms so that the `[some condition ...]` is never satisfied, and I have verified that this code path is never taken.

I was of course expecting the branch to lead to slowdown even if it always picks the `[fast code]` path. But what I was not expecting is that this slowdown depends on the actual `[slow code]`. If I replace it with a slight variation of `[fast code]`, all is fast. But putting in the actual `[slow code]` slows things down (by a factor 2-3), even if this branch is never taken.

Does anyone know why this might happen? Is the branch predictor consistently guessing wrong and executing `[slow code]` only to throw away the result when the condition is found to be false? Anything I can do about this?

**Comment by romainguy** [comment]
Are both code paths computing the same variable(s) used later? It may also be that your slow code path fetches textures without specifying a gradient/LOD, forcing the GPU to move the fetches outside of the condition to compute derivatives.

**Reply by kpeeters** [reply]
The slow path contains multiple new intermediate variable declarations which are not necessary in the fast path. But it's only a few float2, nothing dramatic. It does however reference a static float array, which is unused in the fast path.

Why would the GPU be forced to to do the fetch outside the slow branch when it will never execute that branch? 

(If there's a good place to read about this then feel free to redirect me; I don't want to waste your bandwidth).
